src/poetry/puzzle/provider.py

```diff
@@ -541,7 +555,12 @@ def complete_package(
             if dep.name in self.UNSAFE_PACKAGES:
                 continue
 
-            if self._env and not dep.marker.validate(self._env.marker_env):
+            active_extras = (
+                self._active_root_extras if package.is_root() else dependency.extras
```

Comment:
Seems like the non root part is not tested yet. (Tests succeed when replacing `dependency.extras` with `None`.) Maybe, this path can be triggered by taking one of your tests and introducing an intermediate package?

Reply:
Great catch! While trying to add a test for this I uncovered an issue referenced in a top-level comment. The breaking test I've added doesn't include the intermediate package for simplicity but I can add it as we get to the bottom of that.

Reply:
> that breaking test relied on conflicting versions of the same package between extras which I don't think is a realistic situation that needs to be covered.

Not sure what I was thinking here, that's absolutely an important case in transitive dependencies as well as root (e.g. what if your dependency has a `--cuda` flag to switch Pytorch versions, and you want your own `--cuda` flag to select between the conflicting cpu and cuda versions in the transitive Pytorch dependency).

This was in fact [the bug identified earlier](https://github.com/python-poetry/poetry/pull/9553#pullrequestreview-2341388336). I added a better test based on the earlier one to illustrate it, which covers conflicting dependencies in root and non-root cases. It would pass in the root case (due to duplicate resolution) but fail in the non-root case.

I then **pushed a fix** in [the latest commit](https://github.com/python-poetry/poetry/pull/9553/commits/1326833b925820e2451a7c8a42b5e25c60bb2670), all tests pass but could use validation that it's not unwise: I updated the duplicate resolution logic to _raise `OverrideNeededError`s_ when duplicates conflict even with different extras, but _not perform overlapping marker resolution_ in those "same package with different extras" cases.

I found that in that case overlapping marker resolution results in incomplete lockfiles which don't contain all possible package versions, while the override exceptions are needed to prevent solver failure.

Reply:
> I found that in that case overlapping marker resolution results in incomplete lockfiles which don't contain all possible package versions, while the override exceptions are needed to prevent solver failure.

Yes, overlapping marker resolution is not prepared for duplicates with different extras.

> I updated the duplicate resolution logic to _raise `OverrideNeededError`s_ when duplicates conflict even with different extras, but _not perform overlapping marker resolution_ in those "same package with different extras" cases.

Unfortunately, it is not that simple. The updated logic will only produce correct results if markers are already mutually exclusive. We have to make sure that overrides are not triggered if markers do overlap. I just added a commit with a refined logic. Please take a look.